librad/tests/convergence.rs

```diff
@@ -0,0 +1,40 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::time::Duration;
+
+use tokio::time::timeout;
+
+use librad_test::{logging, rad::testnet};
+
+#[tokio::test]
+async fn converge() -> Result<(), Box<dyn std::error::Error>> {
+    logging::init();
+
+    for num_peers in 1..10 {
+        let peers = testnet::setup(num_peers).await?;
+        timeout(
+            Duration::from_secs(1),
+            testnet::run_on_testnet(peers, num_peers, |mut _apis| async move {
```

Comment:
I'm not sure this actually tests anything :thinking: :joystick:

Interestingly, I'm not even sure if it would make sense to rely on snapshotting the set of currently-connected peers for each peer -- it depends on the network size and the protocol parameters when the network reaches a "steady state". We'll need to make the test harness more modular I think, so we can have e.g. a test for broadcast atomicity (send message, observe all peers eventually receive it) without carrying the git baggage.

Reply:
> I'm not sure this actually tests anything :thinking: :joystick:

Just validates that the counting and boolean logic in `wait_converged` is doing what it should. It was a useful helper while working on the message accumulation.

> Interestingly, I'm not even sure if it would make sense to rely on snapshotting the set of currently-connected peers for each peer -- it depends on the network size and the protocol parameters when the network reaches a "steady state". We'll need to make the test harness more modular I think, so we can have e.g. a test for broadcast atomicity (send message, observe all peers eventually receive it) without carrying the git baggage.

Absolutely. Ideally the outcome is that we have a test suite with different param configs and actually assert the partial view per peer. Would leave that for a future exercise, unless it is critical to accept this PR.

Reply:
#344